lib/internal/modules/run_main.js

```diff
@@ -76,22 +78,27 @@ function shouldUseESMLoader(mainPath) {
   if (mainPath && StringPrototypeEndsWith(mainPath, '.mjs')) { return true; }
   if (!mainPath || StringPrototypeEndsWith(mainPath, '.cjs')) { return false; }
 
-  const { readPackageScope } = require('internal/modules/package_json_reader');
-  const pkg = readPackageScope(mainPath);
-  // No need to guard `pkg` as it can only be an object or `false`.
-  switch (pkg.data?.type) {
-    case 'module':
-      return true;
-    case 'commonjs':
-      return false;
-    default: { // No package.json or no `type` field.
-      if (getOptionValue('--experimental-detect-module')) {
-        // If the first argument of `containsModuleSyntax` is undefined, it will read `mainPath` from the file system.
-        return containsModuleSyntax(undefined, mainPath);
-      }
-      return false;
+  const response = getNearestParentPackageJSONType(mainPath);
+
+  // No package.json or no `type` field.
+  if (response === undefined || response[0] === 'none') {
+    if (getOptionValue('--experimental-detect-module')) {
```

Comment:
`experimental-detect-module` was removed somewhat recently. Will that be backported? If so, does that effect this?